src/proof/arith_proof_recorder.cpp

```diff
@@ -0,0 +1,71 @@
+/*********************                                                        */
+/*! \file arith_proof_recorder.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief A class for saving the skeletons of a arithmetic proofs for later.
+ **/
+
+#include "proof/arith_proof_recorder.h"
+
+#include <algorithm>
+#include <vector>
+
+namespace CVC4 {
+namespace proof {
+
+ArithProofRecorder::ArithProofRecorder() : d_lemmasToFarkasCoefficients()
+{
+  // Nothing else
+}
+void ArithProofRecorder::saveFarkasCoefficients(
+    Node conflict, theory::arith::RationalVectorCP farkasCoefficients)
+{
+  Assert(conflict.getKind() == kind::AND);
+  Assert(conflict.getNumChildren() == farkasCoefficients->size());
+  Debug("pf::arith") << "Saved Farkas Coefficients:" << std::endl;
+  for (size_t i = 0; i < conflict.getNumChildren(); ++i)
+  {
+    const Node& child = conflict[i];
```

Comment:
Can you move the declaration of `child` and `r` before the loop and just do the assignments in the loop?

Reply:
I guess it's a stylistic question.

Since `child` and `r` only have meaning inside the loop, declaring them inside the loop aligns their lexical scopes with their regions of semantic validity.

What would be the advantage of declaring them outside the loop?

Reply:
Won't performance on debug suffer from re-creating these two variables over and over again? Or are we certain this will be solved by the compiler?

Reply:
Yeah, the compiler should lift declarations if doing so is a performance win.

Reply:
Doesn't [this](https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#local-variables) guideline apply here?

Reply:
Hmm, well these variables aren't objects.

Reply:
Yeah, actually, I can't lift them out of the loop anyway, because you can't re-assign references.